docs/proposals/20210305-sidecarset-hotupgrade.md

```diff
@@ -0,0 +1,335 @@
+        SIDECARSET_VERSION_ALT: (v1:metadata.annotations['version_alt.sidecarset.kruise.io/nginx-sidecar-2'])
+```
+The environment variable SIDECARSET_VERSION_ALT records the version of another sidecar container, therefore,
+no longer needs to communicate with each other, in three scenarios:
+
+1. SIDECARSET_VERSION>0, SIDECARSET_VERSION_ALT=0
+
+This scenario indicates that the sidecar is not currently in the hot upgrade process and should perform an initialization operation (e.g., ListenFd)
+
+2. SIDECARSET_VERSION>0, SIDECARSET_VERSION_ALT>0, SIDECARSET_VERSION>SIDECARSET_VERSION_ALT
+
+This scenario indicates that the sidecar is in the process of hot upgrade, and the current version of the sidecar is new, so state migration needs to be performed in the PostStartHook
+
+3. SIDECARSET_VERSION>0, SIDECARSET_VERSION_ALT>0, SIDECARSET_VERSION<SIDECARSET_VERSION_ALT
+
+This scenario indicates that the sidecar is in the process of hot upgrade, and the current version of the sidecar is old, which will only happen if the old sidecar exits abnormally.
+so it should not do any listening or other operations after starting, and just running there without doing anything.
+
+**2. State Migration**
+
+State migration is essentially the migration of stateful resources from the old version to the new version without interrupting service to the outside world. For example, for envoy, it is the migration of fd from the old sidecar to the new sidecar. fd can be passed between processes via Unix Domain Socket, but simply passing fd is still far from the goal of hot upgrade.
+We can divide fd into two types: ListenFd and ConnectionFd, ListenFd is easier to migrate due to its stateless property, while ConnectionFd is not easy to complete the migration, the recommended migration solution is to migrate ListenFd, ConnectionFd "gracefully offline" and re-establish the connection to the new service process at the next request.
+
+**3. Exception Handling**
+
+- **new version of sidecar container keeps failing to start (non-postStartHook execution failure)**: Image, or some other reasons could cause the repeating start-up failure of the new version of sidecar container. In this failure, the status of the pod will be not ready, but the old version of sidecar still remains running.
+under the circumstances, SidecarSet upgrades will be blocked here basing on the MaxUnavailable field unless a newer success version would be released.
+- **failure during state migration (e.g. ListenFd migration succeeded, ConnectionFd graceful offline failed)**: In this case, the state resources (e.g. ListenFd) in the old version of the sidecar container will not be automatically restored, and the kubelet will restart the sidecar container at a later time.
+In this scenario, it is recommended that PostStartHook perform state migration no longer, but performs initialization operations to complete the start of the new version sidecar.
```

Comment:
> In this scenario, it is recommended that PostStartHook perform state migration no longer, but performs initialization operations to complete the start of the new version sidecar.
In this scenario,  PostStartHook MAY skip the state migration, and performs initialization operations to complete the start of the new version sidecar.